pkg/integrations/register.go

```diff
@@ -0,0 +1,240 @@
+package integrations
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+var (
+	registeredIntegrations = []Config{}
+	configFieldNames       = make(map[reflect.Type]string)
+
+	emptyStructType = reflect.TypeOf(struct{}{})
+	configsType     = reflect.TypeOf(Configs{})
+)
+
+// RegisterIntegration dynamically registers a new integration. The Config
+// will represent the configuration that controls the specific integration.
+// Registered Configs may be loaded using UnmarshalYAML or manually
+// constructed.
+//
+// RegisterIntegration panics if cfg is not a pointer.
+func RegisterIntegration(cfg Config) {
+	if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
+		panic(fmt.Sprintf("RegisterIntegration must be given a pointer, got %T", cfg))
+	}
+	registeredIntegrations = append(registeredIntegrations, cfg)
+	configFieldNames[reflect.TypeOf(cfg)] = cfg.Name()
+}
+
+// Configs is a list of integrations.
+type Configs []Config
+
+func (c *Configs) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	return c.unmarshalWithIntegrations(registeredIntegrations, unmarshal)
+}
+
+func (c *Configs) unmarshalWithIntegrations(integrations []Config, unmarshal func(interface{}) error) error {
+	// Create a dynamic struct type full of our registered integrations and
+	// unmarshal to it.
+	var fields []reflect.StructField
+	for _, cfg := range integrations {
+		fields = append(fields, reflect.StructField{
+			Name: "Config_" + cfg.Name(),
+			Tag:  reflect.StructTag(fmt.Sprintf(`yaml:"%s"`, cfg.Name())),
+			Type: reflect.TypeOf(cfg),
+		})
+	}
+
+	var (
+		structType = reflect.StructOf(fields)
+		structVal  = reflect.New(structType)
+	)
+	if err := unmarshal(structVal.Interface()); err != nil {
+		return err
+	}
+
+	// Go over all non-nil fields in structVal and append them to c.
+	structVal = structVal.Elem()
+	for i := 0; i < structVal.NumField(); i++ {
+		if structVal.Field(i).IsNil() {
+			continue
+		}
+
+		val := structVal.Field(i).Interface().(Config)
```

Comment:
Any weird scenario you can think of where this assertion will fail and therefore panic? I know the integrations must satisfy `Config` b/c of the function param.

Reply:
Not within the context we're using this function; our tests should cover it. 

The reflect package is _nasty_ and will panic every time it's used incorrectly, so it's possible to call this function in any number of ways to get a panic if you pass an unexpected struct to it.

Reply:
I guess there's no point trying to handle them, we'd end up exiting the program anyway.